src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.ChecksumToSolutionCache.SolutionAndInFlightCount.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        private sealed partial class ChecksumToSolutionCache
+        {
+            /// <summary>
+            /// Wrapper around asynchronously produced solution.  The computation for producing the solution will be
+            /// canceled when the number of in-flight operations using it goes down to 0.
+            /// </summary>
+            public sealed class SolutionAndInFlightCount
+            {
+                private readonly ChecksumToSolutionCache _cache;
+                private readonly Checksum _solutionChecksum;
+
+                private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+                /// <summary>
+                /// Initially set to 1 to represent the operation that requested and is using this solution.  This also
+                /// allows us to use 0 to represent a point that this solution computation is canceled and can not be
+                /// used again.
+                /// </summary>
+                private int _inFlightCount = 1;
+
+                // For assertion purposes.
+                public int InFlightCount => _inFlightCount;
+
+                public SolutionAndInFlightCount(
+                    ChecksumToSolutionCache cache,
+                    Checksum solutionChecksum,
+                    Func<CancellationToken, Task<Solution>> getSolutionAsync)
+                {
+                    _cache = cache;
+                    _solutionChecksum = solutionChecksum;
+                    Task = getSolutionAsync(_cancellationTokenSource.Token);
+                }
+
+                public Task<Solution> Task { get; }
+
+                public void IncrementInFlightCount()
+                {
+                    using (_cache._gate.DisposableWait(CancellationToken.None))
```

Comment:
takign this gate is never cancellable in this type.  We always want the increment/decrement code (and any cleanup that causes) to always run so we're in a consistent state.